### PR TITLE
Phase class added, which stores the phase of a revolution between [0.0, 1.0)

### PR DESCRIPTION
**Pull request by mitchmickaliger**

### CODE REVIEW

File: src/cpp/pss/astrotypes/units/Phase.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#ifndef PSS_ASTROTYPES_UNITS_PHASE_H
+#define PSS_ASTROTYPES_UNITS_PHASE_H
+
+#include "pss/astrotypes/units/Time.h"
+#include "pss/astrotypes/units/Frequency.h"
+#include "pss/astrotypes/utils/ModuloOne.h"
+#include <boost/units/quantity.hpp>
+#include <boost/units/base_units/angle/degree.hpp>
+#include <boost/units/systems/angle/revolutions.hpp>
+#include <boost/units/conversion.hpp>
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/Phase.h]: Is this include necessary?

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+} // namespace
+
+// Multiplication helpers
+
+// Using the appropriate multiply_typeof_helper below, returns the results
+// of a multiplication operation with the type that matches the templated class X
+template<class Unit, class X>
+inline
+typename multiply_typeof_helper<quantity<Unit, pss::astrotypes::utils::ModuloOne<X>>, X>::type
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/detail/Phase.cpp]: Why is this here and not in ModuloOne.cpp?

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    typedef typename multiply_typeof_helper<X, quantity<Unit, pss::astrotypes::utils::ModuloOne<X>>>::type type;
+
+    return type::from_value(lhs * static_cast<X>(rhs.value()));
+}
+
+// Using the appropriate multiply_typeof_helper below, returns the results
+// of a multiplication operation with the type that matches the templated class X
+template<class Unit, class X>
+inline
+typename multiply_typeof_helper<quantity<Unit, pss::astrotypes::utils::ModuloOne<X>>, X>::type
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/detail/Phase.cpp]: Why is this here and not in ModuloOne.cpp?

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+};
+
+// Division helpers
+
+// Using the appropriate divide_typeof_helper below, returns the results
+// of a division operation with the type that matches the templated class X
+template<class Unit, class X>
+inline
+typename divide_typeof_helper<quantity<Unit, pss::astrotypes::utils::ModuloOne<X>>, X>::type
+operator/(const X& lhs, const quantity<Unit, pss::astrotypes::utils::ModuloOne<X>>& rhs)
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/detail/Phase.cpp]: Shouldn't this also be in ModuloOne.cpp?

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    return type::from_value(lhs/static_cast<X>(rhs.value()));
+}
+
+// Using the appropriate divide_typeof_helper below, returns the results
+// // of a division operation with the type that matches the templated class X
+template<class Unit, class X>
+inline
+typename divide_typeof_helper<quantity<Unit, pss::astrotypes::utils::ModuloOne<X>>, X>::type
+operator/(const quantity<Unit, pss::astrotypes::utils::ModuloOne<X>>& lhs, const X& rhs)
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/detail/Phase.cpp]: Shouldn't this also be in ModuloOne.cpp?

File: src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+{
+}
+
+TEST_F(PhaseTest, test_modulo)
+{
+    Phase<double> p1(1.0 * revolution);
+    Phase<double> p2(0.0 * revolution);
+    ASSERT_EQ(p1, p2);
+}
+
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp:67]: Can we have a short comment for each test just outlining what it does?

File: src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+void PhaseTest::TearDown()
+{
+}
+
+TEST_F(PhaseTest, test_modulo)
+{
+    Phase<double> p1(1.0 * revolution);
+    Phase<double> p2(0.0 * revolution);
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp:61]: Should test that p1.value() returns 0 not 1.

File: src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    auto f = phase/t;
+    static_assert(std::is_same<decltype(f), boost::units::quantity<boost::units::unit<boost::units::frequency_dimension, boost::units::si::system>, double>>::value, "Expecting a frequency type");
+    ASSERT_EQ(phase.value(), f.value());
+}
+
+TEST_F(PhaseTest, test_conversion_from_radians)
+{
+    boost::units::quantity<boost::units::si::plane_angle, utils::ModuloOne<double>> radians(boost::math::constants::pi<double>()/4 * boost::units::si::radians);
+    Phase<double> phase(radians);
+    ASSERT_FLOAT_EQ(phase.value(), Phase<double>(0.125 * revolution).value());
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp]: Only tests case where we have < 1 revolution. What is we have >1 rev? Can a test be added for this?

File: src/cpp/pss/astrotypes/units/Time.h [hunk 9]
```diff
@@ -41,12 +41,16 @@ namespace units {
...
 typedef boost::units::time_dimension Time;
 
-// time points
+// Time points
 typedef JulianClock::time_point JulianDate;
 typedef ModifiedJulianClock::time_point ModifiedJulianDate;
 typedef ModifiedJulianClock::time_point MJD;
 typedef J2000Clock::time_point J2000;
 
+// Determine if the templated type has a time dimension; if so, EnableIfIsTimeUnit will be aliased to type T
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/units/Time.h:51]: Need to add <type_traits> to allow this to work generally.

File: src/cpp/pss/astrotypes/utils/ModuloOne.h [hunk 10]
```diff
@@ -0,0 +1,99 @@
+/*
```
Comment (pulsarben) [inline src/cpp/pss/astrotypes/utils/ModuloOne.h]: We assume that changes here are already covered by the Modulo MR. Integrate with master and let us know :)

**Review by pulsarben — CHANGES_REQUESTED** [review]

File: src/cpp/pss/astrotypes/units/Phase.h (hunk 1 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/Phase.h]: no, it was leftover from previous work; removed. Superfluous boost/units/base_units/angle/degree.hpp include also removed.

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp (hunk 2 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/detail/Phase.cpp]: no good reason; moved to ModuloOne.cpp

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp (hunk 3 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/detail/Phase.cpp]: Moved to ModuloOne.cpp

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp (hunk 4 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/detail/Phase.cpp]: yes, moved

File: src/cpp/pss/astrotypes/units/detail/Phase.cpp (hunk 5 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/detail/Phase.cpp]: yes, moved

File: src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp (hunk 6 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp:67]: Comments added to the top of each test

File: src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp (hunk 7 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp:61]: test added to check that p1 == 0

File: src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp (hunk 8 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/test/src/PhaseTest.cpp]: 3 more tests added, including one where the input is greater than 1 revolution

File: src/cpp/pss/astrotypes/units/Time.h (hunk 9 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/units/Time.h:51]: <type_traits> now #include'd

File: src/cpp/pss/astrotypes/utils/ModuloOne.h (hunk 10 above)
Comment (mitchmickaliger) [reply src/cpp/pss/astrotypes/utils/ModuloOne.h]: integration with master successful after a few conflicts

**Review by pulsarben — APPROVED** [review]

**Pull request merged**